crates/trie/sparse/src/trie.rs

```diff
@@ -1057,7 +1057,193 @@ impl<P> RevealedSparseTrie<P> {
+        // First, do a quick check if the value exists in our values map.
+        // We assume that if there exists a leaf node, then its value will
+        // be in the `values` map.
+        if let Some(actual_value) = self.values.get(path) {
+            // We found the leaf, check if the value matches (if expected value was provided)
+            if let Some(expected) = expected_value {
+                if actual_value != expected {
+                    return Err(LeafLookupError::ValueMismatch {
+                        path: path.clone(),
+                        expected: Some(expected.clone()),
+                        actual: actual_value.clone(),
+                    });
+                }
+            }
+            return Ok(LeafLookup::Found);
+        }
+
+        // If the value does not exist in the `values` map, then this means that the leaf either:
+        // - Does not exist in the trie
+        // - Is missing from the witness
+        // We traverse the trie to find the location where this leaf would have been, showing
+        // that it is not in the trie. Or we find a blinded node, showing that the witness is
+        // not complete.
+        while current.len() < path.len() {
+            match self.nodes.get(&current) {
+                Some(SparseNode::Empty) | None => {
+                    // None implies no node is at the current path (even in the full trie)
+                    // Empty node means there is a node at this path and it is "Empty"
+                    return Ok(LeafLookup::NotFound { diverged_at: current });
+                }
+                Some(&SparseNode::Hash(hash)) => {
+                    // We hit a blinded node - cannot determine if leaf exists
+                    return Err(LeafLookupError::BlindedNode { path: current.clone(), hash });
+                }
+                Some(SparseNode::Leaf { key, .. }) => {
+                    // We found a leaf node before reaching our target depth
+                    let mut full_path = current.clone();
+                    full_path.extend_from_slice_unchecked(key);
+
+                    if &full_path == path {
+                        // TODO: This should be handled by the values map check above
+                        // TODO: But it's here for consistency
+                        if let Some(value) = self.values.get(path) {
+                            if let Some(expected) = expected_value {
+                                if value != expected {
+                                    return Err(LeafLookupError::ValueMismatch {
+                                        path: path.clone(),
+                                        expected: Some(expected.clone()),
+                                        actual: value.clone(),
+                                    });
+                                }
+                            }
+                            return Ok(LeafLookup::Found);
+                        }
+                    }
```

Comment:
this should be unreachable? ah, ye, see the comment. would be great to dedup these

Reply:
yep will dedup